archive/100-theorems-list/16_abel_ruffini.lean

```diff
@@ -0,0 +1,179 @@
+noncomputable def Φ : polynomial R := X ^ 5 - C ↑a * X + C ↑b
+
+variables {R}
+
+@[simp] lemma map_Phi {S : Type*} [comm_ring S] (f : R →+* S) : (Φ R a b).map f = Φ S a b :=
+by simp [Φ]
+
+@[simp] lemma coeff_zero_Phi : (Φ R a b).coeff 0 = ↑b :=
+by simp [Φ, coeff_X_pow]
+
+@[simp] lemma coeff_five_Phi : (Φ R a b).coeff 5 = 1 :=
+by simp [Φ, coeff_X, coeff_C, -C_eq_nat_cast, -ring_hom.map_nat_cast]
+
+variables [nontrivial R]
+
+lemma degree_Phi : (Φ R a b).degree = ↑5 :=
+begin
+  suffices : degree (X ^ 5 - C ↑a * X) = ↑5,
+  { rwa [Φ, degree_add_eq_left_of_degree_lt],
+    rw this,
+    exact lt_of_le_of_lt degree_C_le (with_bot.coe_lt_coe.mpr (nat.zero_lt_bit1 2)) },
+  rw [degree_sub_eq_left_of_degree_lt, degree_X_pow],
+  rw [degree_X_pow, ←pow_one X],
+  exact lt_of_le_of_lt (degree_C_mul_X_pow_le _ _)
+    (with_bot.coe_lt_coe.mpr (nat.one_lt_bit1 two_ne_zero)),
+end
+
+lemma nat_degree_Phi : (Φ R a b).nat_degree = 5 :=
+nat_degree_eq_of_degree_eq_some (degree_Phi a b)
+
+lemma leading_coeff_Phi : (Φ R a b).leading_coeff = 1 :=
+by rw [leading_coeff, nat_degree_Phi, coeff_five_Phi]
+
+lemma monic_Phi : (Φ R a b).monic :=
+leading_coeff_Phi a b
+
+lemma irreducible_Phi (p : ℕ) (hp : p.prime) (hpa : p ∣ a) (hpb : p ∣ b) (hp2b : ¬ (p ^ 2 ∣ b)) :
+  irreducible (Φ ℚ a b) :=
+begin
+  rw [←map_Phi a b (int.cast_ring_hom ℚ), ←is_primitive.int.irreducible_iff_irreducible_map_cast],
+  apply irreducible_of_eisenstein_criterion,
+  { rwa [ideal.span_singleton_prime (int.coe_nat_ne_zero.mpr hp.ne_zero),
+      int.prime_iff_nat_abs_prime] },
+  { rw [leading_coeff_Phi, ideal.mem_span_singleton],
+    exact_mod_cast mt nat.dvd_one.mp (hp.ne_one) },
+  { intros n hn,
+    rw ideal.mem_span_singleton,
+    rw [degree_Phi, with_bot.coe_lt_coe] at hn,
+    interval_cases n with hn;
+    simp [Φ, coeff_X_pow, coeff_C, int.coe_nat_dvd.mpr, hpb, hpa, -ring_hom.eq_int_cast] },
+  { simp only [degree_Phi, ←with_bot.coe_zero, with_bot.coe_lt_coe, nat.succ_pos'] },
+  { rwa [coeff_zero_Phi, pow_two, ideal.span_singleton_mul_span_singleton, ←pow_two,
+      ideal.mem_span_singleton, ←int.coe_nat_pow],
+    norm_num,
+    exact mt int.coe_nat_dvd.mp hp2b },
+  all_goals { exact polynomial.monic.is_primitive (monic_Phi a b) },
+end
+
+lemma real_roots_Phi_le : fintype.card ((Φ ℚ a b).root_set ℝ) ≤ 3 :=
+begin
+  rw [←map_Phi a b (algebra_map ℤ ℚ), Φ, ←one_mul (X ^ 5), ←C_1],
+  refine (card_root_set_le_derivative _).trans
+    (nat.succ_le_succ ((card_root_set_le_derivative _).trans (nat.succ_le_succ _))),
+  suffices : ((C ((algebra_map ℤ ℚ) 20) * X ^ 3).root_set ℝ).subsingleton,
+  { norm_num [fintype.card_le_one_iff_subsingleton, ← mul_assoc, *] at * },
+  rw root_set_C_mul_X_pow; norm_num,
+end
+
+lemma real_roots_Phi_ge_aux (hab : b < a) :
+  ∃ x y : ℝ, x ≠ y ∧ aeval x (Φ ℚ a b) = 0 ∧ aeval y (Φ ℚ a b) = 0 :=
+begin
+  let f := λ x : ℝ, aeval x (Φ ℚ a b),
+  change ∃ x y : ℝ, x ≠ y ∧ f x = 0 ∧ f y = 0,
```

Comment:
This line is unnecessary.
```suggestion
```

Reply:
It was necessary for the change that you mention below, but reverting both is fine.